### PR TITLE
125 multiple language support

### PR DESCRIPTION
**Pull request by Barnemat**
#125 
Adds an extention to the filter, that makes it possible to filter lessons by language. The lanuages are OR  filtrated, which means that lessons from multiple languages may me chosen simultaniously, as opposed to the "tema"-tag, which is AND filtrated.

Some important features:

- When a main language is selected in the language dropdown, the whole filter will be cleared. but with the newly chosen language checked (if the chosen language has any corresponding lessons).

- Added a suggestion for a function that translates group names in the filter (and corresponding translations). At this moment, it only supports the "language" group name.

- Lessons that don't have a language specified in their frontmatter, will be given an "undefined language" tag. This had to be done, because none of the external courses have a defined language, and I didn't want them to be completely hidden. This may be removed, when no longer needed. Lessons that are tagged with an invalid language are ignored.

- Commented in the setLanguage function in the Lesson component, such that when moving to a lesson the language is changed to the lesson language. This was originally meant as a feature for when navigating directly to a lesson by an external link, but I see now that this will also change the language when moving to a lesson from within the page, and therefore might not be desired funtionality. Will redo, if it's undesired.

- Added some selectors, and changes to the getLessons selector, such that the language tags are always shown with the complete number of lessons available, given the constraints in "Tema". Was done to make the language tags "OR-compatible".

- Added a function in util.js that defines the globally valid languages. Updated the places defining valid languages to use this function instead.

- I changed some selectors that still used the getFilteredLessons selector to use the getFilteredAndIndexedLessons selector instead. This removes many lessons (e.g. all "Trysil videokurs"), as many are not indexed. If there's a problem with this, let me know. This was done because I noticed the same problem as described in issue #276 . Many unindexed lessons are lacking important information in their frontmatter (e.g. level), and should, perhaps, not be part of the lesson count. Some indexed lessons are also hidden, as they don't have an assigned level or playlist, e.g. [this lesson](http://oppgaver.kidsakoder.no/beta/scratch/data/data).

### CODE REVIEW

**Comment by Barnemat** [comment]
PS: For testing purposes, it's helpful to reconfigure some of the language tags in the lessons of your "oppgaver" repository, as there should only be the two tags "undefined language" and "norsk bokmål" by default (these are the only categories having corresponding lessons). There will probably also be shown a tag for "english", which corresponds to a readme-file (web/ressurser/readme) not following the naming convention (it's filename is lowercase, but should be uppercase).

File: src/components/Filter/FilterGroup.js [hunk 1]
```diff
@@ -3,24 +3,43 @@ import FilterItem from './FilterItem';
...
+*/
+const getLanguageTagItemName = (tagItem) => {
+  return languages[tagItem] ? languages[tagItem] : tagItem === 'undefined language' ? tagItem : undefined;
+};
+
+/*
+* Extend this function when more groupName translations are added
+*/
+const translateGroupName = (groupName, t) => {
+  return groupName === 'language' ? t('filter.language') : groupName;
```
Comment (krskog) [inline src/components/Filter/FilterGroup.js]: kan vel bare `return t('filter.group.' + groupName)` og lagre grupper i captions som `filter: group: language: "Language"`? Slipper man å gjøre det senere og skreddersyr litt mindre :bowtie:

File: src/constants/captions_da.js [hunk 2]
```diff
@@ -81,6 +81,7 @@ export default {
       textline1: 'I filteret kan man sortere ut de oppgavene man vil løse etter hvilke tema man vil jobbe med.',
       textline2: 'Bak hvert valg står det antall oppgaver som kan løses, etter hvilke valg du gjør i filteret.'
     },
+    language: 'Språk',
```
Comment (krskog) [inline src/constants/captions_da.js]: `filter: group: language: "Språk"`, se første kommentar

File: src/constants/captions_en.js [hunk 3]
```diff
@@ -82,6 +82,7 @@ export default {
       textline2: 'Behind each choice there are the number of lessons that can be solved, according to the choices ' +
                  'you make in the filter.'
     },
+    language: 'Language',
```
Comment (krskog) [inline src/constants/captions_en.js]: ^

File: src/constants/captions_nb.js [hunk 4]
```diff
@@ -81,6 +81,7 @@ export default {
       textline1: 'I filteret kan man sortere ut de oppgavene man vil løse etter hvilke tema man vil jobbe med.',
       textline2: 'Bak hvert valg står det antall oppgaver som kan løses, etter hvilke valg du gjør i filteret.'
     },
+    language: 'Språk',
```
Comment (krskog) [inline src/constants/captions_nb.js]: ^

File: src/constants/captions_nn.js [hunk 5]
```diff
@@ -81,6 +81,7 @@ export default {
       textline1: 'I filteret kan man sortere ut de oppgavene man vil løse etter hvilke tema man vil jobbe med.',
       textline2: 'Bak hvert valg står det antall oppgaver som kan løses, etter hvilke valg du gjør i filteret.'
     },
+    language: 'Språk',
```
Comment (krskog) [inline src/constants/captions_nn.js]: ^

File: src/constants/captions_sv.js [hunk 6]
```diff
@@ -81,6 +81,7 @@ export default {
       textline1: 'I filteret kan man sortere ut de oppgavene man vil løse etter hvilke tema man vil jobbe med.',
       textline2: 'Bak hvert valg står det antall oppgaver som kan løses, etter hvilke valg du gjør i filteret.'
     },
+    language: 'Språk',
```
Comment (krskog) [inline src/constants/captions_sv.js]: ^

File: src/components/Filter/FilterGroup.js [hunk 7]
```diff
@@ -3,24 +3,43 @@ import FilterItem from './FilterItem';
...
 const FilterGroup = React.createClass({
   render(){
-    const groupName = capitalize(this.props.groupName);
+    const groupName = capitalize(translateGroupName(this.props.groupName, this.props.t));
     const filterTags = this.props.tagItems;
     const filterItems = Object.keys(filterTags).map((tagItem) => {
-      const onCheck = () => this.props.onFilterCheck(groupName, tagItem);
+      const onCheck = () => this.props.onFilterCheck(this.props.groupName, tagItem);
       const availableLessonsForTag = this.props.availableLessonsForTag[tagItem];
+      const tagName = this.props.groupName === 'language' ? getLanguageTagItemName(tagItem) : tagItem;
```
Comment (krskog) [inline src/components/Filter/FilterGroup.js]: Dette blir kanskje litt for skreddersydd om du velger å gjøre hva som står i kommentaren over

**Comment by Skagevang** [comment]
Ser bra ut! Men "Kurs på andre nettsider" har blitt borte. Ellers er jeg enig i det Kristian skriver om hvordan det bør være med captions.

File: src/components/Filter/FilterGroup.js [hunk 8]
```diff
@@ -3,24 +3,43 @@ import FilterItem from './FilterItem';
...
+* Extend this function when more groupName translations are added
+*/
+const translateGroupName = (groupName, t) => {
+  return groupName === 'language' ? t('filter.language') : groupName;
+};
 
 const FilterGroup = React.createClass({
   render(){
-    const groupName = capitalize(this.props.groupName);
+    const groupName = capitalize(translateGroupName(this.props.groupName, this.props.t));
```
Comment (NorwegianKiwi) [inline src/components/Filter/FilterGroup.js]: Mangler `t` i propTypes.

File: src/util.js [hunk 9]
```diff
@@ -198,6 +217,16 @@ export function removeHtmlFileEnding(lessonPage) {
 }
 
 /**
+* IMPORTANT:
+* Returns languages defined as available
+* All available languages must be defined here
+* @returns {Object}
+*/
+export const getAvailableLanguages = () => {
+  return {nb: 'norsk bokmål', nn: 'nynorsk', sv: 'svenska', da: 'dansk', en: 'english'};
```
Comment (NorwegianKiwi) [inline src/util.js]: Nå er språk-"navnene" ("Norsk bokmål", "English" etc) oversatt både her og i NavBar.js. Jeg foreslår å flytte dette ut til "captions_xx.js"-filene, eller i det minste bare ha det ett sted. Må gjerne la flagg-definisjonene være samme sted, hvis mulig.

File: src/components/Filter/FilterGroup.js (hunk 1 above)
Comment (NorwegianKiwi) [inline src/components/Filter/FilterGroup.js]: Enig med Kristian. Gruppene skal jo uansett være forhåndsdefinert. Det snakkes om å flytte grupper og tags (inkl. oversettelser) til oppgave-repoet, men dette forslaget er like bra uansett, tror jeg (faktisk tror jeg det vil gjøre det lettere at alt er uniformt og mindre skreddersydd).

File: src/components/Filter/FilterGroup.js [hunk 10]
```diff
@@ -3,24 +3,43 @@ import FilterItem from './FilterItem';
...
+import {getAvailableLanguages} from '../../util';
+
+const languages = getAvailableLanguages();
+
+/*
+* Checks if a tagItem is a valid language tag or if the language tag is 'undefined language'
+* Returns undefined if not.
+*/
+const getLanguageTagItemName = (tagItem) => {
+  return languages[tagItem] ? languages[tagItem] : tagItem === 'undefined language' ? tagItem : undefined;
```
Comment (NorwegianKiwi) [inline src/components/Filter/FilterGroup.js]: Jeg synes vi allerede nå skal fjerne "undefined language", ettersom oppgave-repoet begynner å få språk på plass. Og det gjør det bare enda mer fristende å legge inn språk i tags så fort som mulig.

File: src/components/Navigation/NavBar.js [hunk 11]
```diff
@@ -53,14 +54,31 @@ LanguageItem.propTypes = {
...
+  props.setLanguage(eventKey);
+  const filter = {};
+  Object.keys(props.filter).forEach(group => {
+    filter[group] = {};
+    Object.keys(props.filter[group]).forEach(tag => {
+      if(group === 'language' && tag === eventKey){
+        filter[group][tag] = true;
+      }else{
+        filter[group][tag] = false;
+      }
```
Comment (NorwegianKiwi) [inline src/components/Navigation/NavBar.js]: Denne if-blokken kan byttes ut med `filter[group][tag] = (group === 'language' && tag === eventKey);`

File: src/components/Navigation/NavBar.js [hunk 12]
```diff
@@ -53,14 +54,31 @@ LanguageItem.propTypes = {
...
+    });
+  });
+  props.resetFilter();
```
Comment (NorwegianKiwi) [inline src/components/Navigation/NavBar.js]: I stedet for setFilterAndLanguage, kunne man ikke enten utvidet `resetFilter` litt slik at den tok inn (optionally) én groupName og én tagName (begge kan være undefined), og dersom de er satt, så er det kun den ene kombinasjonen som får true? Se resetFilter i src/reducers/filter.js. Nå gjør denne funksjonen nesten det samme som resetFilters gjør.

File: src/components/Navigation/NavBar.js [hunk 13]
```diff
@@ -53,14 +54,31 @@ LanguageItem.propTypes = {
...
+  props.setFilter(filter);
+};
+
 function LanguageDropdown(props) {
```
Comment (NorwegianKiwi) [inline src/components/Navigation/NavBar.js]: Det virker litt som om veldig mange props blir sendt ned til LanguageDropdown gjennom flere lag som ikke bruker disse props'ene. Og det begynner å bli mange definerte komponenter i NavBar.js. Kanskje man skulle skilt disse ut i litt flere filer? Og samtidig kunne man kanskje hente inn state og/eller actioncreators nærmere der de skal brukes? Jeg pleier å bruke tommelfingerregelen at dersom en state eller actioncreator må sendes "gjennom" en komponent som ikke bruker den selv, så er det bedre at man ikke sender den gjennom, men lar den nederste komponenten selv hente dette inn via mapStateToProps og mapDispatchToProps.

File: src/selectors/lesson.js [hunk 14]
```diff
@@ -40,13 +40,62 @@ export const getFilteredAndIndexedLessons = createSelector(
...
+      for(const tag of Object.keys(filter[groupName])){
+        if(groupName === 'language'){
+          languageFilter[groupName][tag] = true;
+        }
+        else if(filter[groupName][tag]){
+          languageFilter[groupName][tag] = true;
+        }
+        else{
+          languageFilter[groupName][tag] = false;
+        }
```
Comment (NorwegianKiwi) [inline src/selectors/lesson.js]: Her kan det vel også gjøres litt forenklinger. Den siste if-blokken kan feks skrives `languageFilter[groupName][tag] = !!filter[groupName][tag]`.

Dessuten er det vel også spesialbehandling av language her som vi bør prøve å unngå,  slik @krskog nevner over.

Comment (NorwegianKiwi) [inline src/selectors/lesson.js]: Når det er sagt, så begynner disse selectorene å bli litt vel skreddersydde. Tenker vi kan generalisere litt her. Men vi trenger ikke gjøre det i denne PR'en. Synes vi skal lage en issue som heter noe sånn som "Rydde opp i lesson-selectors" der det står at man bør "generalisere" dem eller noe slikt. @Skagevang, gidder du?

File: src/util.js [hunk 15]
```diff
@@ -56,9 +56,19 @@ export function getLessons(lessonContext, readmeContext, courseContext) {
     const courseName = path.slice(2, path.indexOf('/', 2)).toLowerCase();
 
     const courseFrontmatter = courseContext(coursePath).frontmatter;
-    const lessonFrontmatter = lessonContext(path).frontmatter;
+    let lessonFrontmatter = lessonContext(path).frontmatter;
 
-    const language = lessonFrontmatter.language;
+    // The 'undefined language' tag is added because some courses (e.g. external courses)
+    // does not have language defined. May be removed when no longer necessary.
+    const language = lessonFrontmatter.language ? lessonFrontmatter.language : 'undefined language';
```
Comment (NorwegianKiwi) [inline src/util.js]: Som sagt, synes vi skal ta bort undefined language igjen allerede nå.

File: src/util.js [hunk 16]
```diff
@@ -56,9 +56,19 @@ export function getLessons(lessonContext, readmeContext, courseContext) {
...
+
+    if(language){
+      if(lessonFrontmatter.tags){
+        lessonFrontmatter.tags['language'] = language;
+      }else{
+        lessonFrontmatter.tags = {language: [language]};
```
Comment (NorwegianKiwi) [inline src/util.js]: Jeg forstår ikke, hvorfor endrer man på lessonFrontmatter.tags her?

Comment (Barnemat) [inline src/util.js]: Jeg trengte å få dyttet language taggen fra YAML inn under tags, slik at det skulle fungere med filterfunksjonene. Nå er jo language definert utenfor tags i oppgavefilene. 
Eks.
---
language: nb
tags:
  tema: [spill]
---

File: src/components/Filter/FilterGroup.js (hunk 7 above)
Comment (Barnemat) [inline src/components/Filter/FilterGroup.js]: Brukes bare for å få language-taggene til å se fine ut (eks. 'Norsk bokmål' heller enn 'Nb'). Kan evt. slås sammen med en oversettelsesfunksjon for alle filtertaggene, etter hvert.

File: src/components/Filter/FilterGroup.js (hunk 1 above)
Comment (Skagevang) [inline src/components/Filter/FilterGroup.js]: Gruppene generelt for filteret skal være forhåndsdefinert i oppgaverepoet, i tillegg til captions for de. Men spårkdelen til filteret kan ha en fil i src/constants som f.eks. heter filterlanguages.js. Språk i filteret skal stå på den måten det skrives i det landet, språket blir brukt til. F.eks. hvis man har siden på engelsk, skal det fortsatt stå Norsk bokmål i filteret under Language-gruppen. Så filtergruppen for språk er det eneste du trenger å tenke på som skal oversettes.

**Comment by Skagevang** [comment]
Ser tanken med å at hvis man får en link til en engelsk oppgave, så bytter den språk. Men tror dette er forvirrende. Så synes du burde fjerne at den bytter hovedspråket i forhold til oppgaven. Spesielt nå som localStorage er implementert og språket vil bli husket

**Comment by Barnemat** [comment]
Har ordnet en løsning på problemene beskrevet ovenfor. Har lagt til funskjonalitet som gjør det enklere å legge til flere OR-grupper i filteret, men har ikke funnet noen optimal løsning på hvordan selectorene, som har med oppgavetelling å gjøre, skal fungere mellom flere OR-grupper, men selve filtreringen fungerer som den skal.

File: src/components/Navigation/LanguageDropdown.scss [hunk 17]
```diff
@@ -0,0 +1,20 @@
+@import '../../styles/variables';
```
Comment (Skagevang) [inline src/components/Navigation/LanguageDropdown.scss]: Du trenger ikke importere denne hvis du ikke skal bruke noen variabler derfra

File: src/constants/captions_da.js [hunk 18]
```diff
@@ -81,6 +81,10 @@ export default {
       textline1: 'I filteret kan man sortere ut de oppgavene man vil løse etter hvilke tema man vil jobbe med.',
       textline2: 'Bak hvert valg står det antall oppgaver som kan løses, etter hvilke valg du gjør i filteret.'
     },
+    group: {
+      language: 'Språk',
+      tema: 'Tema'
```
Comment (Skagevang) [inline src/constants/captions_da.js]: Du kan fjerne tema taggen, siden dette skal gjøres i et annet issue og alle tags (og oversettelser) skal ligge i oppgaverepo, utenom språk.

File: src/constants/captions_en.js [hunk 19]
```diff
@@ -82,6 +82,10 @@ export default {
       textline2: 'Behind each choice there are the number of lessons that can be solved, according to the choices ' +
                  'you make in the filter.'
     },
+    group: {
+      language: 'Language',
+      tema: 'Topic'
```
Comment (Skagevang) [inline src/constants/captions_en.js]: Se kommenter over

File: src/constants/captions_nb.js [hunk 20]
```diff
@@ -81,6 +81,10 @@ export default {
       textline1: 'I filteret kan man sortere ut de oppgavene man vil løse etter hvilke tema man vil jobbe med.',
       textline2: 'Bak hvert valg står det antall oppgaver som kan løses, etter hvilke valg du gjør i filteret.'
     },
+    group: {
+      language: 'Språk',
+      tema: 'Tema'
```
Comment (Skagevang) [inline src/constants/captions_nb.js]: Se kommenter over

File: src/constants/captions_nn.js [hunk 21]
```diff
@@ -81,6 +81,10 @@ export default {
       textline1: 'I filteret kan man sortere ut de oppgavene man vil løse etter hvilke tema man vil jobbe med.',
       textline2: 'Bak hvert valg står det antall oppgaver som kan løses, etter hvilke valg du gjør i filteret.'
     },
+    group: {
+      language: 'Språk',
+      tema: 'Tema'
```
Comment (Skagevang) [inline src/constants/captions_nn.js]: Se kommentar over

File: src/constants/captions_sv.js [hunk 22]
```diff
@@ -81,6 +81,10 @@ export default {
       textline1: 'I filteret kan man sortere ut de oppgavene man vil løse etter hvilke tema man vil jobbe med.',
       textline2: 'Bak hvert valg står det antall oppgaver som kan løses, etter hvilke valg du gjør i filteret.'
     },
+    group: {
+      language: 'Språk',
+      tema: 'Tema'
```
Comment (Skagevang) [inline src/constants/captions_sv.js]: Se kommentar over

**Comment by Skagevang** [comment]
Virker som all funksjonalitet er bra nå! Noen småting å fikse bare for min del. Kanskje Lance ser andre større ting :)

**Comment by Skagevang** [comment]
Fjern filter fungerer ikke

File: src/action_creators.js [hunk 23]
```diff
@@ -16,9 +16,13 @@ export function setFilter(filter) {
   };
 }
 
-export function resetFilter() {
+export function resetFilter(groupName = {}, tagName = {}) {
   return {
-    type: 'RESET_FILTER'
+    type: 'RESET_FILTER',
+    payload: {
```
Comment (NorwegianKiwi) [inline src/action_creators.js]: Trenger ikke ligge inni "payload"

File: src/action_creators.js [hunk 24]
```diff
@@ -16,9 +16,13 @@ export function setFilter(filter) {
   };
 }
 
-export function resetFilter() {
+export function resetFilter(groupName = {}, tagName = {}) {
```
Comment (NorwegianKiwi) [inline src/action_creators.js]: Trenger vi å sette default-verdier til tomme objekter? Kan det ikke bare står `resetFilter(groupName, tagName)`?

File: src/components/Filter/ActiveFilterItem.js [hunk 25]
```diff
@@ -1,12 +1,14 @@
...
+import {capitalize, getLanguageName} from '../../util';
 import styles from './ActiveFilterItem.scss';
 import withStyles from 'isomorphic-style-loader/lib/withStyles';
 import Button from 'react-bootstrap/lib/Button';
 import Glyphicon from 'react-bootstrap/lib/Glyphicon';
 
 const ActiveFilterItem = (props) => {
-  const tagItem = capitalize(props.tagItem);
+  let tagItem = props.groupName === 'language' ?
+   getLanguageName(props.tagItem) : props.tagItem;
```
Comment (NorwegianKiwi) [inline src/components/Filter/ActiveFilterItem.js]: Dersom vi skal sjekke for `groupName==='language'` spesifikt, så vil jeg at det bare skal gjøres ett sted i hele CC-viewer, ikke både for ActiveFilterItem og for FilterGroups (hvis mulig).

File: src/reducers/language.js [hunk 26]
```diff
@@ -11,7 +12,7 @@ export default function(state='nb', action) {
 }
 
 function isLanguageValid(language) {
-  const languages = ['nb','nn','sv','da','en'];
+  const languages = getAvailableLanguages(true);
```
Comment (NorwegianKiwi) [inline src/reducers/language.js]:
```
function isLanguageValid(language) {
  return getAvailableLanguages().hasOwnProperty(language);
}
```

File: src/util.js [hunk 27]
```diff
@@ -215,6 +247,53 @@ export function removeHtmlFileEnding(lessonPage) {
...
 
 /**
+* IMPORTANT:
+* Returns languages defined as available
+* All available languages must be defined here
+* Dummy argument needed as tests does not allow require statements
+* @param {Dummy argument} noUrl
+* @returns {Array or Object}
+*/
+export const getAvailableLanguages = (onlyKeys) => {
```
Comment (NorwegianKiwi) [inline src/util.js]:
```
export const getAvailableLanguages = () => {
  return {
    'nb': {
      name: 'Norsk bokmål',
      url: require('./assets/graphics/norway.svg')
    },
    'nn': {
      name: 'Norsk nynorsk',
      url: require('./assets/graphics/norway.svg')
    },
    'sv': {
      name: 'Svenska',
      url: require('./assets/graphics/sweden.svg')
    },
    'da': {
      name: 'Dansk',
      url: require('./assets/graphics/denmark.svg')
    },
    'en': {
      name: 'English',
      url: require('./assets/graphics/english.svg')
    }
  };
};
```

Comment (Barnemat) [inline src/util.js]: Grunnen til at det ikke ble gjort sånn i den leverte løsningen var pga. at testene ikke likte "require". Ser ut som det er dette som gjør at Travis krasjer i @Skagevang sin nye PR.

**Comment by Barnemat** [comment]
@Skagevang @NorwegianKiwi Ser at Skagevang har åpnet ny PR som bygger videre på denne. Betyr det at han har tatt over (hadde vært helt OK for min del), og at jeg bør "close" denne PRen?

**Comment by Skagevang** [comment]
@Barnemat Vi tenkte å merge din først også min rett etterpå, så du får creds for jobben ;)

**Comment by Barnemat** [comment]
@Skagevang Ah, kult. Skal jeg bare la den være som den er da, eller?

**Comment by Skagevang** [comment]
Hvis du vil fikse den, så er det bare å gjøre det? Akkurat nå ser jeg på et annet issue. Opp til deg ;) Men si ifra

**Pull request merged**